fail2ban/server/utils.py

```diff
@@ -0,0 +1,199 @@
+			If output is True, the function returns tuple (success, stdoutdata, stderrdata, returncode)
+
+		Returns
+		-------
+		bool
+			True if the command succeeded.
+
+		Raises
+		------
+		OSError
+			If command fails to be executed.
+		RuntimeError
+			If command execution times out.
+		"""
+		stdout = stderr = None
+		retcode = 0
+		try:
+			stime = time.time()
+			popen = subprocess.Popen(
+				realCmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=shell)
+			retcode = popen.poll()
+			while retcode is None and time.time() - stime <= timeout:
+				time.sleep(Utils.DEFAULT_SLEEP_INTERVAL)
+				retcode = popen.poll()
+			if retcode is None:
+				logSys.error("%s -- timed out after %i seconds." %
+					(realCmd, timeout))
+				os.kill(popen.pid, signal.SIGTERM) # Terminate the process
+				time.sleep(Utils.DEFAULT_SLEEP_INTERVAL)
+				retcode = popen.poll()
+				#logSys.debug("%s -- terminated %s ", realCmd, retcode)
+				if retcode is None: # Still going...
+					os.kill(popen.pid, signal.SIGKILL) # Kill the process
+					time.sleep(Utils.DEFAULT_SLEEP_INTERVAL)
+					retcode = popen.poll()
+					#logSys.debug("%s -- killed %s ", realCmd, retcode)
+		except OSError, e:
+			logSys.error("%s -- failed with %s" % (realCmd, e))
+
+		std_level = retcode == 0 and logging.DEBUG or logging.ERROR
+		# if we need output (to return or to log it): 
+		if output or std_level >= logSys.getEffectiveLevel():
+			# if was timeouted (killed/terminated) - to prevent waiting, set std handles to non-blocking mode.
+			if popen.stdout:
+				try:
+					if retcode < 0:
+						Utils.setFBlockMode(popen.stdout, False)
+					stdout = popen.stdout.read()
+				except IOError, e:
+					logSys.error(" ... -- failed to read stdout %s", e)
+				if stdout is not None and stdout != '':
+					logSys.log(std_level, "%s -- stdout: %r", realCmd, stdout)
+				popen.stdout.close()
+			if popen.stderr:
+				try:
+					if retcode < 0:
+						Utils.setFBlockMode(popen.stderr, False)
+					stderr = popen.stderr.read()
+				except IOError, e:
+					logSys.error(" ... -- failed to read stderr %s", e)
+				if stderr is not None and stderr != '':
+					logSys.log(std_level, "%s -- stderr: %r", realCmd, stderr)
+				popen.stderr.close()
+
+		if retcode == 0:
+			logSys.debug("%s -- returned successfully", realCmd)
+			return True if not output else (True, stdout, stderr, retcode)
+		elif retcode is None:
+			logSys.error("%s -- unable to kill PID %i" % (realCmd, popen.pid))
+		elif retcode < 0:
+			logSys.error("%s -- killed with %s" %
+				(realCmd, signame.get(-retcode, "signal %i" % -retcode)))
+		else:
+			msg = _RETCODE_HINTS.get(retcode, None)
+			logSys.error("%s -- returned %i" % (realCmd, retcode))
+			if msg:
+				logSys.info("HINT on %i: %s", retcode, msg % locals())
+			return False if not output else (False, stdout, stderr, retcode)
+		raise RuntimeError("Command execution failed: %s" % realCmd)
+	
+	@staticmethod
+	def wait_for(cond, timeout, interval=None):
```

Comment:
fail2ban uses both styles, this one is imho more speaking as "system" like routine...